wagtail/wagtailcore/models.py

```diff
@@ -107,21 +107,59 @@ def find_for_request(request):
         NB this means that high-numbered ports on an extant hostname may
         still be routed to a different hostname which is set as the default
         """
+
+        # default site q
+        q = Q(is_default_site=True)
+
+        # hostname q
         try:
-            hostname = request.META['HTTP_HOST'].split(':')[0]  # KeyError here goes to the final except clause
-            try:
-                # find a Site matching this specific hostname
-                return Site.objects.get(hostname=hostname)  # Site.DoesNotExist here goes to the final except clause
-            except Site.MultipleObjectsReturned:
-                # as there were more than one, try matching by port too
-                port = request.META['SERVER_PORT']  # KeyError here goes to the final except clause
-                return Site.objects.get(hostname=hostname, port=int(port))
-                # Site.DoesNotExist here goes to the final except clause
-        except (Site.DoesNotExist, KeyError):
-            # If no matching site exists, or request does not specify an HTTP_HOST (which
-            # will often be the case for the Django test client), look for a catch-all Site.
-            # If that fails, let the Site.DoesNotExist propagate back to the caller
-            return Site.objects.get(is_default_site=True)
+            hostname = request.META['HTTP_HOST'].split(':')[0]
+            q |= Q(hostname=hostname)
+        except KeyError:
+            pass
+
+        # get all matching sites
+        # order by is_default_site so if there is a default it is either first or last
+        site_list = list(Site.objects.filter(q).order_by('is_default_site'))
+
+        # only one result (fastpath)
+        if len(site_list) == 1:
+            return site_list[0]
+
+        # DoesNotExist
+        if len(site_list) == 0:
+            raise Site.DoesNotExist()
+
+        # find the default_site (if any)
+        # remove from list if hostname does not match
+        # we need to check both ends as different dbs store and order booleans differently (see django #19726)
+        default_site = None
```

Comment:
If one had thousands of sites, the original 3 query version would be
optimal. This version would be optimal if the count is low, or the total
sites with matching hostname is low.

On 28 December 2015 at 07:18, Karl Hobley <notifications@github.com> wrote:

> In wagtail/wagtailcore/models.py
> <https://github.com/torchbox/wagtail/pull/1963#discussion_r48457952>:
>
> > +        # get all matching sites
> > +        # order by is_default_site so if there is a default it is either first or last
> > +        site_list = list(Site.objects.filter(q).order_by('is_default_site'))
> > +
> > +        # only one result (fastpath)
> > +        if len(site_list) == 1:
> > +            return site_list[0]
> > +
> > +        # DoesNotExist
> > +        if len(site_list) == 0:
> > +            raise Site.DoesNotExist()
> > +
> > +        # find the default_site (if any)
> > +        # remove from list if hostname does not match
> > +        # we need to check both ends as different dbs store and order booleans differently (see django #19726)
> > +        default_site = None
>
> The way I have written it is O(1): no matter how big the site_list, the
> default site will be either first or last (or neither).
>
> I don't think the complexity of this particular filter really matters. For
> most sites, site_list wouldn't have more than a few items. And even if
> one did have thousands of sites, the effort required to pull them all out
> of the database would probably be much greater anyway.
>
> I'm happy for this to be merged without my suggestion though, just thought
> it made the code easier to follow.
>
> but I fail to see negligable impact
>
> Sorry, I meant the list comprehension has a negligable impact
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/torchbox/wagtail/pull/1963/files#r48457952>.
>